### PR TITLE
Implement a UI for generating a new diff

### PR DESCRIPTION
**Pull request by cosletmihai**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -10,4 +10,4 @@ COPY . /usr/src/app
 RUN pip install --upgrade pip
 RUN pip install -r requirements.txt
 
-CMD gunicorn --bind 0.0.0.0:${RDF_DIFFER_API_PORT} rdf_differ.entrypoints.flask_app:app
\ No newline at end of file
+CMD gunicorn --bind 0.0.0.0:${RDF_DIFFER_API_PORT} rdf_differ.entrypoints.api.run:app
```
Comment (costezki) [inline Dockerfile:13]: I wonder how to run more than one service in a single Docker container.

File: rdf_differ/config.py [hunk 2]
```diff
@@ -14,3 +14,5 @@
 
 FUSEKI_USERNAME = os.environ.get('FUSEKI_USERNAME', 'admin')
 FUSEKI_PASSWORD = os.environ.get('FUSEKI_PASSWORD', 'admin')
+
+FLASK_SECRET_KEY = os.environ.get('SECRET_KEY', 'secret key')
```
Comment (costezki) [inline rdf_differ/config.py:18]: what is this the key for?

File: rdf_differ/entrypoints/api/handlers.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ def get_diffs() -> tuple:
     fuseki_adapter = FusekiDiffAdapter(config.ENDPOINT, http_client=requests, sparql_client=SPARQLRunner())
     try:
         datasets = fuseki_adapter.list_datasets()
-        return [{dataset: fuseki_adapter.dataset_description(dataset)} for dataset in datasets], 200
+        return [fuseki_adapter.dataset_description(dataset) for dataset in datasets], 200
```
Comment (costezki) [inline rdf_differ/entrypoints/api/handlers.py:29]: looks cleaner

File: rdf_differ/entrypoints/ui/api.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/python3
+
+# api.py
+# Date:  18/09/2020
+# Author: Mihai Coșleț
+# Email: coslet.mihai@gmail.com
+
+"""
+Service to consume API
```
Comment (costezki) [inline rdf_differ/entrypoints/ui/api.py:9]: I think the file name is misleading, a small documentation will help as well, explain the purpose of this file.

File: rdf_differ/entrypoints/ui/config.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+# Author: Mihai Coșleț
+# Email: coslet.mihai@gmail.com
+
+"""
+Module description
+
+"""
+import os
+
+API_ENDPOINT = os.environ.get('API_ENDPOINT', 'http://rdf_differ:3040')
```
Comment (costezki) [inline rdf_differ/entrypoints/ui/config.py:14]: should we call it RDF differ API ? to be explicit.

File: requirements/dev.txt [hunk 6]
```diff
@@ -13,3 +14,10 @@ flake8
 myst-parser
 sphinx<3
 sphinxcontrib-apidoc
+
+
+# Flask requirements for testing
+Flask
+flask-wtf
+flask-bootstrap
+Flask-Wtf
```
Comment (costezki) [inline requirements/dev.txt:23]: Flask-wtf is mentioned twice, is this correct?

File: ui/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+COPY . /usr/src/app
+
+RUN pip install --upgrade pip
+RUN pip install -r requirements.txt
+
+CMD flask run --host=0.0.0.0 --port=${UI_PORT}
```
Comment (costezki) [inline ui/Dockerfile:13]: let's run both of them with gunicorn

File: ui/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+requests
+
+Flask
+flask-wtf
+flask-bootstrap
+Flask-Wtf
```
Comment (costezki) [inline ui/requirements.txt:6]: flask-wtf is mentioned twice.

**Review by costezki — CHANGES_REQUESTED** [review]
Well done! there are, however, a few things you shall revise.

File: Dockerfile (hunk 1 above)
Comment (cosletmihai) [reply Dockerfile:13]: it is certainly possible: [docker documentation](https://docs.docker.com/config/containers/multi-service_container/), [medium article](https://towardsdatascience.com/run-multiple-services-in-single-docker-container-using-supervisor-b2ed53e3d1c0)
but the examples that I saw seem to expose only one service to the "outside" (a **django** service and **redis**) and it I think that the configuration of the endpoints of having both the `api` and the `ui` on the same docker might become problematic (`api.meaningfy.co` and `meaningfy.co`)

File: rdf_differ/config.py (hunk 2 above)
Comment (cosletmihai) [reply rdf_differ/config.py:18]: [docs](https://flask.palletsprojects.com/en/1.1.x/config/#SECRET_KEY):
> A secret key that will be used for securely signing the session cookie and can be used for any other security related needs by extensions or your application. It should be a long random string of bytes, although unicode is accepted too.

File: ui/Dockerfile (hunk 7 above)
Comment (cosletmihai) [reply ui/Dockerfile:13]: it's not recommended to run gunicorn without having a nginix in front of it, and most of the articles I read don't use gunicorn for development purposes, and it also doesn't forward code errors (see attached picture) which will make development more cumbersome, thus I don't think it's a good idea to use gunicorn for development.

error1

**Review by costezki — APPROVED** [review]
Well Done! Thank You!

**Pull request merged**